[CLD-22]: feat(operation): introduce retry mechanism on operation

graham-chainlink opened the pull request:
Introducing default retry policy, `no retry` option and `custom input hook` option which allows user to customize the input before retrying eg for customizing gas limit.

Note: Report idempotency behaviour will be introduced in future PR

JIRA: https://smartcontract-it.atlassian.net/browse/CLD-22

cgruber commented on deployment/operations/execute.go:
```diff
@@ -0,0 +1,82 @@
+// ExecuteOp executes an operation with the given input and dependencies.
+// By default, it retries the operation up to 10 times with exponential backoff if it fails.
+// Use WithNewInputHook and WithNoRetry to customize the retry behavior.
+func ExecuteOp[Input, Output, Deps any](
```
If this is in operations, does it need to say Op? As in, operations.Execute(a, b, c) seems really clear.

graham-chainlink replied:
This is because we have sequence too, so eventually we will have `ExecuteOp` (will rename to `ExecuteOperation`) and also `ExecuteSequence`

cgruber commented on deployment/operations/execute.go:
```diff
@@ -0,0 +1,82 @@
+	noRetry   bool
+}
+
+type ExecuteOption[Input any] func(*ExecuteConfig[Input])
```
I really like this.

cgruber commented on deployment/operations/execute.go:
```diff
@@ -0,0 +1,82 @@
+	operation *Operation[Input, Output, Deps],
+	deps Deps,
+	input Input,
+	opts ...ExecuteOption[Input],
```
I like the flexibility, but it doesn't protect the developer very well from using multiple, incompatible options. The logic flow does seem to handle it at the moment, because of the no-retry if statement, but it means the control is implicit, and if a user provided both, they'd get no signal they'd made an error in the usage.

I would suggest that these, despite my liking generalization, the current options all relate to retry policy (none, or custom handler). I'd consider not having "Opts" but rather having "retryPolicy" and using this current code, renamed to just be about retries. IF we have more general options, we can evolve it there.

cgruber replied:
The mechanism is fine, it's mostly the use of variadic opts, and calling them general opts, rather than RetryPolicy. The config approach is great.

cgruber replied:
Or actually... are you using variadics to allow someone to say `ExecuteOp(e, o, in, out)` or `ExecuteOp(e, o, in, out, opt)`?  If that's the case, and it's just to sort of simulate function overloading in a way, then I guess that's fine, but I would document the crap about that usage pattern, and raise an error if someone adds in more than one.

graham-chainlink replied:
Good point on the retry policy rather than Opts, so what i did now is that i consolidated all the retry related config into a single option `WithRetryConfig`, have a look and let me know what you think

graham-chainlink committed:
feat(operation): introduce retry mechanism on operation

Introducing default retry policy, no retry option and custom input hook option which allows user to customize the input before retrying eg for customizing gas limit.

Note: Report idempotency behaviour will be introduced in future PR

JIRA: https://smartcontract-it.atlassian.net/browse/CLD-22

jkongie commented on deployment/operations/execute.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// WithRetryConfig is an ExecuteOption that sets the retry configuration.
+func WithRetryConfig[IN, DEP any](config RetryConfig[IN, DEP]) ExecuteOption[IN, DEP] {
```
```suggestion
func WithRetryPolicy[IN, DEP any](config RetryConfig[IN, DEP]) ExecuteOption[IN, DEP] {
```

graham-chainlink replied:
I think retry config makes more sense, because the retry is already enabled by default, and we just configuring the settings related to retry (disable or the input hook)

graham-chainlink replied:
policy would mean we introduce different retry strategy which is not the case here.

graham-chainlink committed:
test(operation): extra empty input test

Add extra operation test with empty input

graham-chainlink committed:
test(operation): add error reporter test for execute

graham-chainlink commented on deployment/changeset.go:
```diff
@@ -109,6 +111,7 @@ type ChangesetOutput struct {
 	DescribedTimelockProposals []string
 	MCMSProposals              []mcms.Proposal
 	AddressBook                AddressBook
+	Reports                    []operations.Report[any, any]
```
This allows the operations API to send the reports back to the CLD side for processing

graham-chainlink commented on deployment/environment.go:
```diff
@@ -107,6 +109,7 @@ type Environment struct {
 	Offchain          OffchainClient
 	GetContext        func() context.Context
 	OCRSecrets        OCRSecrets
+	OperationsBundle  operations.Bundle
```
This is the link from CLD to chainlink/deployment, CLD will setup the `bundle` eg like reporter and it will be used in the changsets.

jkongie replied:
We should put a comment in here to explain what this is used for

graham-chainlink replied:
Updated!

graham-chainlink commented on deployment/environment.go:
```diff
@@ -130,6 +133,8 @@ func NewEnvironment(
 		GetContext:        ctx,
 		OCRSecrets:        secrets,
+		// default to memory reporter as that is the only reporter available for now
+		OperationsBundle: operations.NewBundle(ctx, logger, operations.NewMemoryReporter()),
```
This is only used for testing in `chainlink/deployment`
